torchvision/ops/giou_loss.py

```diff
@@ -1,14 +1,9 @@
 import torch
-from torch import Tensor
 
 from ..utils import _log_api_usage_once
+from ._utils import _upcast, _loss_inter_union
 
-
-def _upcast(t: Tensor) -> Tensor:
-    # Protects from numerical overflows in multiplications by upcasting to the equivalent higher type
-    if t.dtype not in (torch.float32, torch.float64):
-        return t.float()
-    return t
```

Comment:
This is not the same `_upcast` as `_utils`. It doesn't support integers and converts everything to floats. Could you please review all the places where the `giou_loss._upcast()` was used and ensure the output will be a float? Basically box ops are OK to maintain things as integers but not losses.

Reply:
Yes I do understand that I have not used the int only _upcast I have intentionally not taken it back. Since I would like to know why 

> box ops are OK to maintain things as integers but not losses.

Reply:
The `_upcast` on boxes was introduced for operators that estimate information related to a box. Things like the area for instance. If you put integer boxes and you ask for the area of a box, you kind of expecting you receive the value in integers (that's debatable but that's how the operator worked). The issue was that if you used too small of a precision, the area estimation would overflow. So this method upcasts the boxes to a space where it's safe to do multiplications without risking overflowing for most applications.

On the losses side, I'm not aware of any application that does things on integer space. Not only that but doing reduce=mean will break things. So we need to be careful to definitely not support ints in losses. Similar care might be needed for some box operators. Area might still make sense to return as integer but I'm not 100% sure if that's the case with all the IoU metrics we deal here.

So I think it's important prior merging this PR, to make an explicit decision of what has to support integers and what doesn't, handle it appropriately and add tests and xfails to ensure we are not breaking anything.